src/kernel.c

```diff
@@ -0,0 +1,642 @@
+#include <libdragon.h>
+#include "kernel.h"
+#include "kernelinternal.h"
+#include "timer.h"
+#include <assert.h>
+#include <stdlib.h>
+#include <memory.h>
+
+#define DEBUG_KERNEL   0
+
+#define STACK_COOKIE   0xDEADBEEFBAADC0DE
+#define STACK_GUARD    64
+
+/** @brief Read the current value of the gp register */
+#define REG_GP()     ({ uint64_t gp; __asm("move %0, $28": "=r" (gp)); gp; })
+
+/**
+ * @brief Execute a context switch.
+ *
+ * This forces a context switch to happen, by invoking the syscall 0x1. The
+ * syscall will be intercepted by the generic interrupt handler (inthandler.S)
+ * that will call #__kthread_syscall_schedule.
+ *
+ * This macro is very low level and is called as part of higher-level primitives
+ * that force a context switch like #thread_yield.
+ *
+ * Do not call this under interrupt; use #KTHREAD_SWITCH_ISR instead.
+ */
+#define KTHREAD_SWITCH()      __asm("syscall 0x1")
+
+/**
+ * @brief Executed a context switch, during an ongoing interrupt service routing.
+ *
+ * This forces a context switch to happen just like #KTHREAD_SWITCH, but it is
+ * the right version to call if the code is currently servicing an interrupt.
+ */
+#define KTHREAD_SWITCH_ISR()  ({ __isr_force_schedule = true; })
```

Comment:
We could in theory, but it would mean we could suspend a thread while it's under interrupt. I'm not sure of the ramification of this. I like the idea that when an interrupt fires, it is being served right away; the fact that somewhere in the interrupt service routine, somebody might switch away from it and the service routine might be suspended for an indefinite period of time seems to break the basic assumption that interrupts are super high priority, even higher than the highest thread. It looks dangerous.

Also, there's a small performance/memory benefit to get, as `KTHREAD_SWITCH_ISR` doesn't cause a second exception but causes the current exception to be used for thread switching as soon as the interrupt is done with. Nothing incredible, but still sounds like a easy gain. 

In general, this might make thing a little bit more complicated in there kernel itself, but it should be mostly transparent to users.

For comparison, FreeRTOS for instance has many duplicated APIs in this same way (normal and ISR version). Nintendo kernel got away with the problem by not allowing to write interrupt service routine in the first place (interrupts are just used to switch threads and fire events, business logic is only in threads); it's maybe a easier design but surely less efficients. I like how we are able to do both with this PR (eg: keeping the AI fed via ISR seems like a good design, irrespective of the fact that you might have a audio thread to handle playback).

Reply:
This totally make sense! Also I don't think it is something we cannot change if need be as it is completely internal.